src/ua/main.cpp

```diff
@@ -529,6 +543,39 @@ void printEnvironmentInfo() {
   }
 }
 
+// There is currently the possibility of a race condition if a chunk
+// upload timed-out.  It's possible that a second upload succeeds,
+// has the chunk marked as "complete" and then the first request makes
+// its way through the queue and marks the chunk as pending again.
+// Since we are just about to close the file, we'll check to see if any
+// chunks are marked as pending, and if so, we'll retry them.
+void check_for_complete_chunks(vector<File> &files) {
+  for (int currCheckNum=0; currCheckNum < NUM_CHUNK_CHECKS; ++currCheckNum){
+    while (!chunksFinished.empty()) {
+      Chunk *c = chunksFinished.consume();
+      if (!is_chunk_complete(c)) {
+          chunksToUpload.produce(c);
+      }
+    }
+    // All of the chunks were marked as complete, so let's exit and we
+    // should be safeish to close the file.
+    if(chunksToUpload.size() == 0)
```

Comment:
Ah, I should read more carefully. I missed the creation & join() of uploadThreads.

What if there's a problem in `uploadChunks`, `uploaded==false`, we don't take the branch at 213? Assuming we take the branch at 224, we'll add a chunk to `chunksToRead` at 239. But the read thread(s) aren't running so nothing further will happen. 
We might need to call `createWorkerThreads` and `joinWorkerThreads` to ensure all necessary threads are able to 'do their thing'.